code/game/objects/items/devices/gps.dm

```diff
@@ -1,28 +1,423 @@
-	to_chat(user, "<span class='notice'>[html_icon(src)] \The [src] flashes <i>[get_coordinates()]</i>.</span>")
+	var/screen_color = "#d98736"
+	var/decal_icon = 'icons/obj/items/device/locator_overlays.dmi'
+	var/gps_tag = "GEN0"
+	var/emped = FALSE
+	var/tracking = FALSE		      // Will not show other signals or emit its own signal if false.
+	var/long_range = FALSE		      // If true, can see farther, depending on get_map_levels().
+	var/local_mode = FALSE		      // If true, only GPS signals of the same Z level are shown.
+	var/hide_signal = FALSE		      // If true, signal is not visible to other GPS devices.
+	var/can_hide_signal = FALSE       // If it can toggle the above var.
+	var/is_special_gps_marker = FALSE // How the GPS marker should be handled.
 
-/obj/item/gps/examine(mob/user)
+	var/mob/holder
+	var/is_in_processing_list = FALSE
+	var/list/tracking_devices
+	var/list/showing_tracked_names
+	var/obj/compass_holder/compass
+	var/list/decals
+
+// Mapped type for away site GPS.
+/obj/item/gps/marker
+	is_special_gps_marker = TRUE
+
+/obj/item/gps/Initialize()
+	global.all_gps_units += src
 	. = ..()
-	to_chat(user, "<span class='notice'>\The [src]'s screen shows: <i>[get_coordinates()]</i>.</span>")
+	name = "[initial(name)] ([gps_tag])"
+	compass = new(src)
+	update_holder()
+	update_icon()
+
+/obj/item/gps/examine(mob/user, distance)
+	. = ..()
+	if(distance <= 1)
+		to_chat(user, SPAN_NOTICE("\The [src]'s screen shows: <i>[get_coordinates()]</i>."))
 
 /obj/item/gps/proc/get_coordinates()
 	var/turf/T = get_turf(src)
 	return T ? "[T.x]:[T.y]:[T.z]" : "N/A"
 
+/obj/item/gps/proc/check_visible_to_holder()
+	. = holder && (src in holder.get_held_items())
+
+/obj/item/gps/proc/update_holder()
+
+	var/decl/observ/moved/moved_event = GET_DECL(/decl/observ/moved)
+	var/decl/observ/dir_set/dir_set_event = GET_DECL(/decl/observ/dir_set)
+
+	if(holder && loc != holder)
+		moved_event.unregister(holder, src)
+		dir_set_event.unregister(holder, src)
+		holder.client?.screen -= compass
+		holder = null
+
+	if(istype(loc, /mob))
+		holder = loc
+		moved_event.register(holder, src, .proc/update_compass)
+		dir_set_event.register(holder, src, .proc/update_compass)
+
+	if(holder && tracking)
+		if(!is_in_processing_list)
+			START_PROCESSING(SSobj, src)
+			is_in_processing_list = TRUE
+		if(holder.client)
+			if(check_visible_to_holder())
+				holder.client.screen |= compass
+			else
+				holder.client.screen -= compass
+	else
+		STOP_PROCESSING(SSobj, src)
+		is_in_processing_list = FALSE
+		if(holder?.client)
+			holder.client.screen -= compass
+
+/obj/item/gps/pickup()
+	. = ..()
+	update_holder()
+
+/obj/item/gps/equipped_robot()
+	. = ..()
+	update_holder()
+
+/obj/item/gps/equipped()
+	. = ..()
+	update_holder()
+
+/obj/item/gps/dropped()
+	. = ..()
+	update_holder()
+
+/obj/item/gps/Process()
+	if(!tracking)
+		is_in_processing_list = FALSE
+		return PROCESS_KILL
+	update_holder()
+	if(holder)
+		update_compass(TRUE)
```

Comment:
Causing me pain, but maybe it's superior to purely using events. If the waypoints don't move, why not just do this with a moved event?

I will note that if you did do this purely with events, you'd need to subscribe to the mob login (and possibly logout) events as well.

Reply:
Waypoints do move, they are generally other GPS units. I thought a periodic update would scale better than having every GPS subscribe to every other GPS.
Why would I need to hook login/logout for the mob? Other than adding the compass to screen.

Reply:
When the client logs back in, things are not automatically placed back on the screen. This block here will probably trigger within 2 seconds anyway and do that, so it's not the biggest deal, but I would probably listen for the login event anyway. The logout event is less clear; you'd only need that if you wanted to do cleanup, like e.g. exit out of compass on logout.

Reply:
I think I am going to leave the `process()` loop and not hook the client login event, unless you feel strongly about it. Registering to other GPS in particular is going to be a complicated thing to implement.